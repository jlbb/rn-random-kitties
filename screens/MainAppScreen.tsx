import React, { Ref, useRef, useEffect } from "react";
import {
  StyleSheet,
  View,
  ScrollView,
  Platform,
  StatusBar,
  RefreshControl
} from "react-native";
import { vh } from "react-native-expo-viewport-units";

import { StickyNavigationBar, ButtonRK, Card } from "../components";
import { useContextBgColor, useCatsApi } from "../store/hooks";
import { scrollToRef } from "../utils/utils";
import { colors } from "../config/contants";

interface MainAppScreenProps {
  nCards?: number;
}

const MainAppScreen = ({ nCards = 3 }: MainAppScreenProps): JSX.Element => {
  const [catItems, fetchRandomCats] = useCatsApi();
  const [bgColor, resetBgColor, setRandomBgColor] = useContextBgColor();
  const scrollViewRef: Ref<ScrollView> = useRef();
  const cardRef: Ref<typeof Card>[] = Array.apply(null, {
    length: nCards
  }).map(() => useRef());
  const [refreshing, setRefreshing] = React.useState(false);

  const cardTitles: ArrayNLength<string, 3> = ["Intro", "About", "Customise!"];

  useEffect(() => {
    fetchRandomCats(nCards);
  }, []);

  const onRefresh = React.useCallback(() => {
    setRefreshing(true);
    fetchRandomCats();

    setTimeout(() => setRefreshing(false), 1500);
  }, [refreshing]);

  const navigateToCard = (cardId): any => {
    cardRef[cardId] && scrollToRef(scrollViewRef, cardRef[cardId], cardId);
  };

  const _renderNavigationBar = (): JSX.Element => (
    <StickyNavigationBar>
      {cardTitles.map((title, key) => (
        <ButtonRK
          key={`key-buttonrk-${key}`}
          title={title}
          onButtonPress={() => navigateToCard(key)}
        />
      ))}
    </StickyNavigationBar>
  );

  const _renderCardList = (catsImg: string[]): JSX.Element[] => {
    const cardText = [
      "Lorem ipsum sin dolor, consectetur adipiscing elit.",
      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. "
    ];

    return catsImg.map((item: string, key: number) => {
      if (key === 2) {
        return (
          <Card
            key={`key-cardlist-${key}`}
            ref={cardRef[key]}
            imageSrc={item}
            title={cardTitles[key]}
            controls
            controlsItem={[
              {
                label: "Reset",
                action: resetBgColor
              },
              {
                label: "Randomise",
                action: setRandomBgColor
              }
            ]}
          />
        );
      }

      return (
        <Card
          key={`key-cardlist-${key}`}
          ref={cardRef[key]}
          imageSrc={item}
          title={cardTitles[key]}
        >
          {cardText[key]}
        </Card>
      );
    });
  };

  return (
    <View style={styles.mainContainer}>
      <View style={[styles.container, { backgroundColor: bgColor }]}>
        {_renderNavigationBar()}
        <ScrollView
          ref={scrollViewRef}
          style={styles.scrollView}
          contentContainerStyle={styles.scrollViewContent}
          refreshControl={
            <RefreshControl
              refreshing={refreshing}
              onRefresh={onRefresh}
              style={styles.refresh}
            />
          }
        >
          {_renderCardList(catItems)}
        </ScrollView>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  mainContainer: {
    flex: 1,
    marginTop: Platform.OS === "ios" ? 0 : StatusBar.currentHeight,
    width: "100%"
  },
  container: {
    flex: 1,
    alignItems: "center",
    backgroundColor: colors.BACKGROUND_COLOR,
    justifyContent: "center",
    paddingTop: 70
  },
  refresh: {},
  scrollView: {
    flex: 1,
    height: vh(90),
    width: "100%"
  },
  scrollViewContent: {
    alignItems: "center",
    paddingBottom: vh(75)
  }
});

export default MainAppScreen;
