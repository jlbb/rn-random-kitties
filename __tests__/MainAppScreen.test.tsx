import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";

import MainAppScreen from "../screens/MainAppScreen";

(global as any).fetch = require("jest-fetch-mock");

const mockedStore = {
  context: {
    backgroundColor: ""
  },
  app: {
    catItems: []
  }
};

const mockStore = configureMockStore([]);

describe("MainAppScreen", () => {
  let store;
  let wrapper;

  beforeAll(() => {
    store = mockStore(mockedStore);
    wrapper = mount(
      <Provider store={store}>
        <MainAppScreen />
      </Provider>
    );
  });

  it("renders correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("renders with 3 cards by default", () => {
    expect(wrapper.find("Card").length).toBe(0);

    (fetch as any).mockResponseOnce([
      "catPic1.jpg",
      "catPic2.jpg",
      "catPic3.jpg"
    ]);
    fetch("*").then(res => {
      expect(res).toEqual(["catPic1.jpg", "catPic2.jpg", "catPic3.jpg"]);
      expect(wrapper.find("Card").length).toBe(3);
      expect(wrapper.find("Card").find("ButtonRK").length).toBe(2);
    });
  });
});
