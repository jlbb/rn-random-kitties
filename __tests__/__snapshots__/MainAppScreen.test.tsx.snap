// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainAppScreen renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MainAppScreen>
    <View
      style={
        Object {
          "flex": 1,
          "marginTop": 0,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "marginTop": 0,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "backgroundColor": "#808080",
                "flex": 1,
                "justifyContent": "center",
                "paddingTop": 70,
              },
              Object {
                "backgroundColor": "",
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#808080",
                  "flex": 1,
                  "justifyContent": "center",
                  "paddingTop": 70,
                },
                Object {
                  "backgroundColor": "",
                },
              ]
            }
          >
            <StickyNavigationBar>
              <View
                style={
                  Object {
                    "backgroundColor": "#ffffff",
                    "flexDirection": "row",
                    "height": 70,
                    "position": "absolute",
                    "top": 0,
                    "width": "100%",
                    "zIndex": 2,
                  }
                }
              >
                <View
                  style={
                    Object {
                      "backgroundColor": "#ffffff",
                      "flexDirection": "row",
                      "height": 70,
                      "position": "absolute",
                      "top": 0,
                      "width": "100%",
                      "zIndex": 2,
                    }
                  }
                >
                  <ButtonRK
                    key="key-buttonrk-0"
                    onButtonPress={[Function]}
                    title="Intro"
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                          }
                        }
                      >
                        <TouchableHighlight
                          activeOpacity={0.85}
                          delayPressOut={100}
                          onPress={[Function]}
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "backgroundColor": "#984fa0",
                                "flexDirection": "row",
                                "height": "100%",
                                "justifyContent": "center",
                                "minHeight": 50,
                              },
                              Object {
                                "backgroundColor": "#984fa0",
                              },
                            ]
                          }
                          underlayColor="black"
                        >
                          <View
                            accessible={true}
                            isTVSelectable={true}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "backgroundColor": "#984fa0",
                                  "flexDirection": "row",
                                  "height": "100%",
                                  "justifyContent": "center",
                                  "minHeight": 50,
                                },
                                Object {
                                  "backgroundColor": "#984fa0",
                                },
                              ]
                            }
                          >
                            <View
                              accessible={true}
                              isTVSelectable={true}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Array [
                                  Object {
                                    "alignItems": "center",
                                    "backgroundColor": "#984fa0",
                                    "flexDirection": "row",
                                    "height": "100%",
                                    "justifyContent": "center",
                                    "minHeight": 50,
                                  },
                                  Object {
                                    "backgroundColor": "#984fa0",
                                  },
                                ]
                              }
                            >
                              <Text
                                style={
                                  Array [
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 18,
                                      "textTransform": "uppercase",
                                    },
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 16,
                                    },
                                  ]
                                }
                              >
                                <Text
                                  style={
                                    Array [
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 18,
                                        "textTransform": "uppercase",
                                      },
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 16,
                                      },
                                    ]
                                  }
                                >
                                  Intro
                                </Text>
                              </Text>
                            </View>
                          </View>
                        </TouchableHighlight>
                      </View>
                    </View>
                  </ButtonRK>
                  <ButtonRK
                    key="key-buttonrk-1"
                    onButtonPress={[Function]}
                    title="About"
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                          }
                        }
                      >
                        <TouchableHighlight
                          activeOpacity={0.85}
                          delayPressOut={100}
                          onPress={[Function]}
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "backgroundColor": "#984fa0",
                                "flexDirection": "row",
                                "height": "100%",
                                "justifyContent": "center",
                                "minHeight": 50,
                              },
                              Object {
                                "backgroundColor": "#984fa0",
                              },
                            ]
                          }
                          underlayColor="black"
                        >
                          <View
                            accessible={true}
                            isTVSelectable={true}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "backgroundColor": "#984fa0",
                                  "flexDirection": "row",
                                  "height": "100%",
                                  "justifyContent": "center",
                                  "minHeight": 50,
                                },
                                Object {
                                  "backgroundColor": "#984fa0",
                                },
                              ]
                            }
                          >
                            <View
                              accessible={true}
                              isTVSelectable={true}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Array [
                                  Object {
                                    "alignItems": "center",
                                    "backgroundColor": "#984fa0",
                                    "flexDirection": "row",
                                    "height": "100%",
                                    "justifyContent": "center",
                                    "minHeight": 50,
                                  },
                                  Object {
                                    "backgroundColor": "#984fa0",
                                  },
                                ]
                              }
                            >
                              <Text
                                style={
                                  Array [
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 18,
                                      "textTransform": "uppercase",
                                    },
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 16,
                                    },
                                  ]
                                }
                              >
                                <Text
                                  style={
                                    Array [
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 18,
                                        "textTransform": "uppercase",
                                      },
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 16,
                                      },
                                    ]
                                  }
                                >
                                  About
                                </Text>
                              </Text>
                            </View>
                          </View>
                        </TouchableHighlight>
                      </View>
                    </View>
                  </ButtonRK>
                  <ButtonRK
                    key="key-buttonrk-2"
                    onButtonPress={[Function]}
                    title="Customise!"
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                          }
                        }
                      >
                        <TouchableHighlight
                          activeOpacity={0.85}
                          delayPressOut={100}
                          onPress={[Function]}
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "backgroundColor": "#984fa0",
                                "flexDirection": "row",
                                "height": "100%",
                                "justifyContent": "center",
                                "minHeight": 50,
                              },
                              Object {
                                "backgroundColor": "#984fa0",
                              },
                            ]
                          }
                          underlayColor="black"
                        >
                          <View
                            accessible={true}
                            isTVSelectable={true}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "backgroundColor": "#984fa0",
                                  "flexDirection": "row",
                                  "height": "100%",
                                  "justifyContent": "center",
                                  "minHeight": 50,
                                },
                                Object {
                                  "backgroundColor": "#984fa0",
                                },
                              ]
                            }
                          >
                            <View
                              accessible={true}
                              isTVSelectable={true}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Array [
                                  Object {
                                    "alignItems": "center",
                                    "backgroundColor": "#984fa0",
                                    "flexDirection": "row",
                                    "height": "100%",
                                    "justifyContent": "center",
                                    "minHeight": 50,
                                  },
                                  Object {
                                    "backgroundColor": "#984fa0",
                                  },
                                ]
                              }
                            >
                              <Text
                                style={
                                  Array [
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 18,
                                      "textTransform": "uppercase",
                                    },
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 16,
                                    },
                                  ]
                                }
                              >
                                <Text
                                  style={
                                    Array [
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 18,
                                        "textTransform": "uppercase",
                                      },
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 16,
                                      },
                                    ]
                                  }
                                >
                                  Customise!
                                </Text>
                              </Text>
                            </View>
                          </View>
                        </TouchableHighlight>
                      </View>
                    </View>
                  </ButtonRK>
                </View>
              </View>
            </StickyNavigationBar>
            <ScrollViewMock
              contentContainerStyle={
                Object {
                  "alignItems": "center",
                  "paddingBottom": 1000.5,
                }
              }
              refreshControl={
                <RefreshControlMock
                  onRefresh={[Function]}
                  refreshing={false}
                  style={Object {}}
                />
              }
              style={
                Object {
                  "flex": 1,
                  "height": 1200.6000000000001,
                  "width": "100%",
                }
              }
            >
              <_class
                contentContainerStyle={
                  Object {
                    "alignItems": "center",
                    "paddingBottom": 1000.5,
                  }
                }
                refreshControl={
                  <RefreshControlMock
                    onRefresh={[Function]}
                    refreshing={false}
                    style={Object {}}
                  />
                }
                style={
                  Object {
                    "flex": 1,
                    "height": 1200.6000000000001,
                    "width": "100%",
                  }
                }
              >
                <RCTScrollView
                  contentContainerStyle={
                    Object {
                      "alignItems": "center",
                      "paddingBottom": 1000.5,
                    }
                  }
                  refreshControl={
                    <RefreshControlMock
                      onRefresh={[Function]}
                      refreshing={false}
                      style={Object {}}
                    />
                  }
                  style={
                    Object {
                      "flex": 1,
                      "height": 1200.6000000000001,
                      "width": "100%",
                    }
                  }
                >
                  <RefreshControlMock
                    onRefresh={[Function]}
                    refreshing={false}
                    style={Object {}}
                  >
                    <_class>
                      <RCTRefreshControl />
                    </_class>
                  </RefreshControlMock>
                  <View>
                    <View />
                  </View>
                </RCTScrollView>
              </_class>
            </ScrollViewMock>
          </View>
        </View>
      </View>
    </View>
  </MainAppScreen>
</Provider>
`;
